dev-docs/RESEARCH_ENFORCEMENT_AND_MECHANICS.md

```
@@ -0,0 +1,336 @@
+}
+
+pub fn parse_log_line(line: &[u8], format: &Format) -> ParsedLine<'_> { /* ... */ }
+```
+
+**PR 5 — OTLP split.** Same pattern: define `OtlpLogRecord` in `logfwd-kernel/src/otlp_wire.rs` for the decoded wire format, move protobuf decoding there, leave `RecordBatch` assembly in logfwd-core. The struct approach (vs. a trait-based `ParseSink`) is simpler and produces values that proptest oracles can directly inspect.
+
+During migration, **existing fuzz targets and proptest tests continue working** through re-exports. Once all five PRs land and the kernel is stable, fuzz targets testing pure parsing logic can optionally move to the kernel crate; tests exercising Arrow integration stay in logfwd-core. The `http` crate (pure types, no IO) + `hyper` (client/server with IO) is the canonical example of this pattern in the Rust ecosystem.
+
+## Conclusion
+
+The kernel crate's guarantees rest on four reinforcing layers. First, `#![no_std]` structurally prevents IO access — no lint to circumvent, no `#[allow]` escape hatch. Second, `forbid(unsafe_code)` plus strict clippy restriction lints catch panics, unwraps, and debug artifacts at compile time. Third, Kani proofs exhaustively verify function contracts against all possible inputs, with bolero providing a unified interface that also serves proptest and fuzzing. Fourth, CI enforces that every public function has a proof, coverage stays at 100%, dependencies don't grow unchecked, and cargo-vet ensures every dependency is audited. These layers are not redundant — each catches failures the others miss. The `no_std` enforcement catches IO that lints might miss; Kani catches logical errors that tests might miss; mutation testing catches assertions that Kani's property specifications might not cover. Together they make the kernel crate the most trustworthy component in the system.
```

markdownlint-cli2 (0.22.0)

[warning] Files should end with a single newline character

(MD047, single-trailing-newline)